web/src/bootstrap_typeahead.ts

```diff
@@ -373,7 +375,29 @@ export class Typeahead<ItemType extends string | object> {
                 // We expect the typeahead creator to handle when to hide / show the typeahead.
                 trigger: "manual",
                 arrow: false,
-                offset: [0, 2],
+                offset({placement, reference}) {
+                    if (input_element.type === "textarea" && input_element.$element[0]) {
+                        const caret = getCaretCoordinates(
+                            input_element.$element[0],
+                            input_element.$element[0].selectionStart,
+                        );
+                        // Used to consider the scroll height of textbox in the vertical offset.
+                        const scrollTop = input_element.$element.scrollTop() ?? 0;
+                        if (placement === "top-start") {
+                            // Gap separates the typeahead and caret by 4px.
+                            const gap = 4;
+                            return [caret.left, -caret.top + scrollTop + gap];
+                        }
+                        // In bottom-start, the offset is calculated from bottom of the popper reference.
```

Comment:
Can you add a blank line before the comment here, and the previous block?

It helps to visually separate out the distinct thoughts.

An alternative, that might be nicer, is to rework this as an `if/else` block with an assert in the second block that it's `bottom-start`, since those are the only two placements declared above, and move these comments inside the block where we've already checked what the placement is.

Reply:
Using `if/else` is not possible since popper on initialization, calculates offsets for all possible placements:

![image](https://github.com/zulip/zulip/assets/78212328/cf18f227-4bdb-47d4-ad6f-182556211721)